File: BasicToAdvance/Java/Day-15/BSTPredecessor.java
public class BSTPredecessor {

    static class Node {
        int val;
        Node left, right;

        Node(int val) {
            this.val = val;
        }
    }

    static Node findMax(Node root) {
        while (root.right != null) {
            root = root.right;
        }
        return root;
    }

    static Node predecessor(Node root, int key) {
        Node predecessor = null;

        while (root != null) {
            if (key > root.val) {
                predecessor = root;
                root = root.right;
            } else if (key < root.val) {
                root = root.left;
            } else {
                break;
            }
        }

        if (root != null && root.left != null) {
            return findMax(root.left);
        }

        return predecessor;
    }

    public static void main(String[] args) {
        Node root = new Node(50);
        root.left = new Node(30);
        root.right = new Node(70);
        root.left.left = new Node(20);
        root.left.right = new Node(40);
        root.right.left = new Node(60);
        root.right.right = new Node(80);

        int key = 60;
        Node result = predecessor(root, key);

        if (result != null) {
            System.out.println("Predecessor of " + key + " is " + result.val);
        } else {
            System.out.println("No predecessor exists for " + key);
        }
    }
}

File: BasicToAdvance/Python/Day-15/BSTPredecessor.py
# BST Predecessor

class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None


def find_max(root):
    while root.right:
        root = root.right
    return root


def predecessor(root, key):
    predecessor = None

    while root:
        if key > root.val:
            predecessor = root
            root = root.right
        elif key < root.val:
            root = root.left
        else:
            break

    if root and root.left:
        return find_max(root.left)

    return predecessor


# Example usage
root = Node(50)
root.left = Node(30)
root.right = Node(70)
root.left.left = Node(20)
root.left.right = Node(40)
root.right.left = Node(60)
root.right.right = Node(80)
print(predecessor(root, 60).val)  # Output: 50
print(predecessor(root, 30).val)  # Output: 20